File: src/main/java/akostenko/aicars/drawing/CarImg.java
package akostenko.aicars.drawing;

import static akostenko.aicars.model.CarModel.axleTrack;
import static akostenko.aicars.model.CarModel.tyreRadius;
import static akostenko.aicars.model.CarModel.wheelbase;
import static java.lang.StrictMath.PI;
import static java.util.stream.Collectors.toList;

import akostenko.aicars.math.Decart;
import akostenko.aicars.math.Vector;
import akostenko.aicars.race.car.Car;

import org.newdawn.slick.Color;

import java.util.Collection;

public class CarImg {

    private static final Vector FL_wheel = new Decart(wheelbase/2, axleTrack/2);
    private static final Vector FR_wheel = new Decart(wheelbase/2, -axleTrack/2);

    private static final Vector RL_wheel_p1 = new Decart(-wheelbase/2 - tyreRadius, axleTrack/2);
    private static final Vector RL_wheel_p2 = new Decart(-wheelbase/2 + tyreRadius, axleTrack/2);
    private static final Vector RR_wheel_p1 = new Decart(-wheelbase/2 - tyreRadius, -axleTrack/2);
    private static final Vector RR_wheel_p2 = new Decart(-wheelbase/2 + tyreRadius, -axleTrack/2);
    private static final Vector rearAxle_p1 = new Decart(-wheelbase/2, axleTrack/2);
    private static final Vector rearAxle_p2 = new Decart(-wheelbase/2, -axleTrack/2);

    private static final Vector carAxis_p1 = new Decart(wheelbase/2, 0);
    private static final Vector carAxis_p2 = new Decart(-wheelbase/2, 0);

    public static Collection<Line> get(Car<?> car, Decart cameraPosition, Color color, Scale scale) {
        Decart screenPosition = car.getPosition().multi(scale.getPixels() / scale.getMeters())
                .plus(cameraPosition);


        return new LinesBuilder()
                .from(FL_wheel).towards(car.getSteering().toPolar().d, tyreRadius)
                .from(FL_wheel).towards(car.getSteering().toPolar().d+PI, tyreRadius)
                .between(FL_wheel, FR_wheel)
                .from(FR_wheel).towards(car.getSteering().toPolar().d, tyreRadius)
                .from(FR_wheel).towards(car.getSteering().toPolar().d+PI, tyreRadius)

                .between(RL_wheel_p1, RL_wheel_p2)
                .between(RR_wheel_p1, RR_wheel_p2)
                .between(rearAxle_p1, rearAxle_p2)

                .between(carAxis_p1, carAxis_p2)
                .build().stream()
                .map(line -> line.rotate(car.getHeading().toPolar().d))
                .map(line -> line.scale(scale))
                .map(line -> line.position(screenPosition, color))
                .collect(toList());
    }

}

File: src/main/kotlin/akostenko/aicars/drawing/CarImg.kt
package akostenko.aicars.drawing

import akostenko.math.vector.Decart
import akostenko.math.vector.Decart.Companion.ZERO
import akostenko.aicars.model.CarModel.axleTrack
import akostenko.aicars.model.CarModel.frontWeightPercent
import akostenko.aicars.model.CarModel.rearWeightPercent
import akostenko.aicars.model.CarModel.tyreRadius
import akostenko.aicars.model.CarModel.wheelbase
import akostenko.aicars.race.car.Car
import akostenko.math.Line
import akostenko.math.StraightLine
import java.lang.Math.PI

object CarImg {

    private val FL_wheel = Decart(wheelbase * frontWeightPercent, axleTrack / 2)
    private val FR_wheel = Decart(wheelbase * frontWeightPercent, -axleTrack / 2)

    private val RL_wheel_p1 = Decart(-wheelbase * rearWeightPercent - tyreRadius, axleTrack / 2)
    private val RL_wheel_p2 = Decart(-wheelbase * rearWeightPercent + tyreRadius, axleTrack / 2)
    private val RR_wheel_p1 = Decart(-wheelbase * rearWeightPercent - tyreRadius, -axleTrack / 2)
    private val RR_wheel_p2 = Decart(-wheelbase * rearWeightPercent + tyreRadius, -axleTrack / 2)
    private val rearAxle_p1 = Decart(-wheelbase * rearWeightPercent, axleTrack / 2)
    private val rearAxle_p2 = Decart(-wheelbase * rearWeightPercent, -axleTrack / 2)

    private val carAxis_p1 = Decart(wheelbase * frontWeightPercent * 1.5, 0.0)
    private val carAxis_p2 = Decart(-wheelbase * rearWeightPercent, 0.0)

    fun build(car: Car<*>): Collection<Line> {

        return StraightLinesBuilder(true)
                .from(FL_wheel).towards(car.steering.toPolar().d, tyreRadius)
                .from(FL_wheel).towards(car.steering.toPolar().d + PI, tyreRadius)
                .between(FL_wheel, FR_wheel)
                .from(FR_wheel).towards(car.steering.toPolar().d, tyreRadius)
                .from(FR_wheel).towards(car.steering.toPolar().d + PI, tyreRadius)

                .between(RL_wheel_p1, RL_wheel_p2)
                .between(RR_wheel_p1, RR_wheel_p2)
                .between(rearAxle_p1, rearAxle_p2)

                .between(carAxis_p1, carAxis_p2)
                .build()
                .map { line -> line.rotate(car.heading.toPolar().d, ZERO) }
                .map { (from, to, collidable) -> StraightLine(from + car.position, to + car.position, collidable) }

    }

}
